feat: uploading to prefix.dev

wolfv opened the pull request.

0xbe7a commented on src/upload.rs:
```diff
@@ -70,3 +94,107 @@ pub async fn upload_package_to_quetz(
+        println!("Upload successful!");
+    } else {
+        println!("Upload failed!");
+        if response.status() == 401 {
```
You can use https://docs.rs/reqwest/latest/reqwest/struct.Response.html#method.error_for_status

wolfv commented:
@0xbe7a I refactored the code a little bit and added progress bars to each of the methods. Do you think you can test this for quetz & artifactory?

This also allows to put more than a single package for uploading, e.g.

`rattler-build upload prefix --channel my-test *.conda`

0xbe7a commented on src/upload.rs:
```diff
@@ -115,37 +131,150 @@ pub async fn upload_package_to_artifactory(
+        let index_json = IndexJson::from_package_directory(package_dir.path()).into_diagnostic()?;
+        let subdir = index_json
+            .subdir
+            .ok_or_else(|| miette::miette!("index.json of the package has no subdirectory. Cannot determine which directory to upload to"))?;
```
We should probably include the package name in the error message

0xbe7a commented:
Looks good, i will test the Quetz and Artifactory upload later today

wolfv commented:
OK, let me know how it goes and then I'll merge :)

0xbe7a commented:
Works as advertised, ready to merge! 🤠
